### PR TITLE
Enable FhirToDatalake pipeline to support FHIR r5

### PR DESCRIPTION
**Pull request by QuanWanxx**
Support FHIR R5

### CODE REVIEW

File: FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/Fhir/SpecificationProviders/Stu3FhirSpecificationProvider.cs [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+        }
+
+        public override bool IsValidFhirResourceType(string resourceType)
+        {
+            return Stu3FhirModelInfo.IsKnownResource(resourceType);
+        }
+
+        public override IEnumerable<string> GetCompartmentResourceTypes(string compartmentType)
+        {
+            throw new NotImplementedException();
```
Comment (sowu880) [inline FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/Fhir/SpecificationProviders/Stu3FhirSpecificationProvider.cs:37]: GetCompartmentResourceTypes and GetSearchParametersByResourceType can be removed.

File: FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/PipelineRegistrationExtensions.cs [hunk 2]
```diff
@@ -34,16 +35,14 @@ public static class PipelineRegistrationExtensions
...
             services.AddSingleton<IGroupMemberExtractor, GroupMemberExtractor>();
 
             services.AddSingleton<ITypeFilterParser, TypeFilterParser>();
 
             services.AddSingleton<IReferenceParser, R4ReferenceParser>();
 
-            services.AddSchemaConverters();
-            return services;
+            services.AddFhirSpecificationProviders();
+            services.AddSchemaConverters();            return services;
```
Comment (sowu880) [inline FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/PipelineRegistrationExtensions.cs:45]: new line.

File: FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/Fhir/SpecificationProviders/BaseFhirSpecificationProvider.cs [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+        /// </summary>
+        private readonly Dictionary<string, HashSet<string>> _compartmentResourceTypesLookup;
+
+        /// <summary>
+        /// The FHIR server supported search parameters for each resource type, extracted from FHIR server metadata.
+        /// </summary>
+        private readonly Dictionary<string, HashSet<string>> _resourceTypeSearchParametersLookup;
+
+        /// <summary>
+        /// {resourceType}_{searchParameter} to search parameter id defined by
```
[the remark this answers is not in the document]
Comment (QuanWanxx) [reply FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/Fhir/SpecificationProviders/BaseFhirSpecificationProvider.cs:39]: Just copy that from current [R4FhirSpecificationProvider](https://github.com/microsoft/FHIR-Analytics-Pipelines/blob/main/FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/Fhir/R4FhirSpecificationProvider.cs#L52)...

File: FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/Fhir/SpecificationProviders/R4FhirSpecificationProvider.cs [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+            {
+                compartment = parser.Parse<CompartmentDefinition>(compartmentContext);
+            }
+            catch (Exception exception)
+            {
+                _logger.LogError($"Failed to parse compartment definition from file {compartmentFile}. Reason: {exception.Message}");
+                throw new FhirSpecificationProviderException($"Failed to parse compartment definition from file {compartmentFile}.", exception);
+            }
+
+            var compartmentType = compartment.Code?.ToString();
```
[the remark this answers is not in the document]
Comment (QuanWanxx) [reply FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/Fhir/SpecificationProviders/R4FhirSpecificationProvider.cs:79]: I feel not, the **compartment** in different fhir versions are in different namespaces, even though they have same name and same methods.

File: FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/Fhir/SpecificationProviders/R4FhirSpecificationProvider.cs [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+            {
+                capabilityStatement = parser.Parse<CapabilityStatement>(metaData);
+            }
+            catch (Exception exception)
+            {
+                _logger.LogError($"Failed to parse capability statement from FHIR server metadata. Reason: {exception.Message}");
+                throw new FhirSpecificationProviderException($"Failed to parse capability statement from FHIR server metadata.", exception);
+            }
+
+            var searchParameters = new Dictionary<string, HashSet<string>>();
```
[the remark this answers is not in the document]
Comment (QuanWanxx) [reply FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/Fhir/SpecificationProviders/R4FhirSpecificationProvider.cs:116]: Also, the **capabilityStatement** for different Fhir versions are in different namespaces, hence they are totally different objects

File: FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/Fhir/SpecificationProviders/R4FhirSpecificationProvider.cs [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+            _logger.LogInformation($"Build SearchParametersLookup from fhir server metadata successfully.");
+
+            return new Tuple<Dictionary<string, HashSet<string>>, Dictionary<string, string>>(searchParameters, searchParameterIds);
+        }
+
+        /// <summary>
+        /// build SearchParameterDefinitionLookup, search parameter url to searchParameter object
+        /// </summary>
+        /// <returns>search parameter url to searchParameter object dictionary</returns>
+        private Dictionary<string, SearchParameter> BuildSearchParameterDefinitionLookup()
```
[the remark this answers is not in the document]
Comment (QuanWanxx) [reply FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/Fhir/SpecificationProviders/R4FhirSpecificationProvider.cs:159]: Also, copy from current [R4FhirSpecificationProvider](https://github.com/microsoft/FHIR-Analytics-Pipelines/blob/main/FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/Fhir/R4FhirSpecificationProvider.cs#L251), and seems be used in constructor, but be commented out

File: FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/Fhir/SpecificationProviders/Stu3FhirSpecificationProvider.cs (hunk 1 above)
Comment (QuanWanxx) [reply FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/Fhir/SpecificationProviders/Stu3FhirSpecificationProvider.cs]: Sure

File: FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/PipelineRegistrationExtensions.cs (hunk 2 above)
Comment (QuanWanxx) [reply FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/PipelineRegistrationExtensions.cs]: Fixed, thanks.

File: FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Common/Configurations/ConfigurationConstants.cs [hunk 7]
```diff
@@ -31,5 +32,8 @@ public static class ConfigurationConstants
 
         // Reference docker's image name format: https://docs.docker.com/engine/reference/commandline/tag/#extended-description
         public static readonly Regex ImageNameRegex = new Regex(@"^[a-z0-9]+(([_\.]|_{2}|\-+)[a-z0-9]+)*(\/[a-z0-9]+(([_\.]|_{2}|\-+)[a-z0-9]+)*)*$");
+
+        // Currently the pipeline support FHIR R4 and R5 schemas.
+        public static readonly List<FhirVersion> SupportedFhirVersions = new List<FhirVersion> { FhirVersion.R4, FhirVersion.R5 };
```
[the remark this answers is not in the document]
Comment (QuanWanxx) [reply FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Common/Configurations/ConfigurationConstants.cs]: Updated, thanks.

File: FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/PipelineRegistrationExtensions.cs [hunk 8]
```diff
@@ -64,5 +65,25 @@ public static IServiceCollection AddSchemaConverters(this IServiceCollection ser
 
             return services;
         }
+
+        public static IServiceCollection AddFhirSpecificationProviders(this IServiceCollection services)
+        {
+            services.AddSingleton<IFhirSpecificationProvider, R4FhirSpecificationProvider>();
+
+            services.AddTransient<R4FhirSpecificationProvider>();
```
Comment (ACMoretxj) [inline FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/PipelineRegistrationExtensions.cs:73]: What's the consideration to mark the specification provider as `Transient`? Seems it won't change at runtime.

File: FhirToDataLake/test/Microsoft.Health.Fhir.Synapse.Core.UnitTests/TestUtils.cs [hunk 9]
```diff
@@ -99,5 +107,23 @@ public static IDataSchemaConverter TestDataSchemaConverterDelegate(string name)
...
+
+        public static IFhirSpecificationProvider TestFhirSpecificationProviderDelegate(FhirVersion fhirVersion)
+        {
+            var dataClient = Substitute.For<IFhirDataClient>();
+
+            var metadataOptions = new MetadataOptions();
+            dataClient.Search(metadataOptions)
+                .ReturnsForAnyArgs(x => TestDataProvider.GetBundleFromFile(TestDataConstants.R4MetadataFile));
+
+            if (fhirVersion == FhirVersion.R4)
```
Comment (ACMoretxj) [inline FhirToDataLake/test/Microsoft.Health.Fhir.Synapse.Core.UnitTests/TestUtils.cs:119]: The `if` and `else` act identically, is it by design?

File: FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/PipelineRegistrationExtensions.cs (hunk 8 above)
Comment (QuanWanxx) [reply FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/PipelineRegistrationExtensions.cs:73]: Fixed, thanks.

File: FhirToDataLake/test/Microsoft.Health.Fhir.Synapse.Core.UnitTests/TestUtils.cs (hunk 9 above)
Comment (QuanWanxx) [reply FhirToDataLake/test/Microsoft.Health.Fhir.Synapse.Core.UnitTests/TestUtils.cs:119]: This is just for test, in actual delegate we use **switch** and will throw exceptions if fhirVersion is undefined.

File: FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/Microsoft.Health.Fhir.Synapse.Core.csproj [hunk 10]
```diff
@@ -9,10 +9,13 @@
     <PackageReference Include="Microsoft.Extensions.Hosting" Version="6.0.0" />
     <PackageReference Include="NCrontab.Signed" Version="3.3.2" />
     <PackageReference Include="Newtonsoft.Json" Version="13.0.1" />
-	<PackageReference Include="Hl7.Fhir.R4" Version="3.5.0">
+    <PackageReference Include="Hl7.Fhir.R5" Version="4.2.1">
+        <Aliases>FhirR5</Aliases>
+    </PackageReference>
+	<PackageReference Include="Hl7.Fhir.R4" Version="4.2.1">
```
[the remark this answers is not in the document]
Comment (QuanWanxx) [reply FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/Microsoft.Health.Fhir.Synapse.Core.csproj]: Fixed, thanks.

File: FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/PipelineRegistrationExtensions.cs [hunk 11]
```diff
@@ -82,5 +82,25 @@ public static IServiceCollection AddSchemaConverters(this IServiceCollection ser
 
             return services;
         }
+
+        public static IServiceCollection AddFhirSpecificationProviders(this IServiceCollection services)
+        {
+            services.AddSingleton<IFhirSpecificationProvider, R4FhirSpecificationProvider>();
+
+            services.AddSingleton<R4FhirSpecificationProvider>();
+            services.AddSingleton<R5FhirSpecificationProvider>();
```
[the remark this answers is not in the document]
Comment (QuanWanxx) [reply FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/PipelineRegistrationExtensions.cs]: Sure, updated to register **SpecificationProvider** based on **FhirVersion**, thanks.

**Review by moria97 — APPROVED** [review]

**Pull request merged**